### DISCUSSION TITLE
Intel oneDAL FATAL ERROR on Windows10

### DISCUSSION THREAD

**Discussion by CHDev93** [intel/scikit-learn-intelex#531] [Q&A] [answered]
**Describe the bug**
The windows wheel is unable to load the dll's needed to run the examples. I tried one of the examples and got the following when calling a library function

Intel oneDAL FATAL ERROR: onedal_thread.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: onedal_sequential.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: Cannot load neither onedal_thread.1.dll nor onedal_sequential.1.dll.
Intel oneDAL FATAL ERROR: onedal_thread.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: onedal_sequential.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: Cannot load neither onedal_thread.1.dll nor onedal_sequential.1.dll.

**To Reproduce**
1. `pip install daal4py pandas lightgbm` on Windows10
2. Alter the `__init__.py` to have `path_to_libs = os.path.join(os.path.dirname(sys.executable), "..\\Library\\bin")` (note the `..`)
3. Run the following script
```python
import lightgbm as lgb
import numpy as np
import daal4py as d4p


def get_data(n, m):
    x_train = np.random.randn(n, m).astype(np.float32)
    A = np.random.randint(-5, 5, size=(m, 1))
    y_train = (x_train @ A).astype(np.float32)

    return x_train, y_train

n = 1000
m = 25
x_train, y_train = get_data(n, m)

params = {
    'task': 'train',
    'boosting_type': 'gbdt',
    'objective': 'regression',
    'metric': ['rmse'],
    'device': 'cpu',
    'num_leaves': 31,
    'bagging_fraction': 0.5,
    'feature_fraction': 0.5,
    'learning_rate': 0.001,
    'verbose': 2,
    'max_bin': 255,
}
ds_train = lgb.Dataset(x_train, y_train.ravel())#, free_raw_data=False)
gbm = lgb.train(
    params,
    ds_train,
    num_boost_round=10,
    # keep_training_booster=args.keep_training_booster,
)

print("Converting...", flush=True)
daal_model = d4p.get_gbt_model_from_lightgbm(gbm)
print("Converted...", flush=True)
daal_prediction = d4p.gbt_regression_prediction().compute(x_train, daal_model).prediction
```

3. See output
```
[LightGBM] [Debug] Dataset::GetMultiBinFromAllFeatures: sparse rate 0.000000
[LightGBM] [Debug] init for col-wise cost 0.000021 seconds, init for row-wise cost 0.000844 seconds
[LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.001624 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 6375
[LightGBM] [Info] Number of data points in the train set: 1000, number of used features: 25
[LightGBM] [Info] Start training from score 0.386410
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 8
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 9
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 9
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 9
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 9
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 6
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 7
[LightGBM] [Debug] Trained a tree with leaves = 31 and max_depth = 8
Converting...
Converted...
Intel oneDAL FATAL ERROR: onedal_thread.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: onedal_sequential.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: Cannot load neither onedal_thread.1.dll nor onedal_sequential.1.dll.
Intel oneDAL FATAL ERROR: onedal_thread.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: onedal_sequential.1.dll. Error code is 0x80096005.
Intel oneDAL FATAL ERROR: Cannot load neither onedal_thread.1.dll nor onedal_sequential.1.dll.
```

**Expected behavior**
Successfully converted lightgbm model

**Output/Screenshots**
Would note that I do see the dlls in the correct folder and the directory is added to PATH

**Environment:**
 - OS: Windows10
 - Compiler: ?
 - Version: ?
 - Python: 3.6.6

**Comment by PetrovKP** [comment]
@CHDev93  This is not a problem with finding a library. Otherwise there would be errors:
```
Intel oneDAL FATAL ERROR: Cannot find/load library onedal_thread.1.dll.
Intel oneDAL FATAL ERROR: Cannot find/load library onedal_sequential.1.dll.
Intel oneDAL FATAL ERROR: Cannot load neither onedal_thread.1.dll nor onedal_sequential.1.dll.
Intel oneDAL FATAL ERROR: Cannot find/load library onedal_thread.1.dll.
Intel oneDAL FATAL ERROR: Cannot find/load library onedal_sequential.1.dll.
Intel oneDAL FATAL ERROR: Cannot load neither onedal_thread.1.dll nor onedal_sequential.1.dll.
```

Your error is more exotic ([from the forum](https://answers.microsoft.com/en-us/protect/forum/protect_defender-protect_start-windows_10/error-0x80096005/b973f073-ec8f-475e-a543-9cc453d7bdbc)): 
```
0x80096005 -2146869243 TRUST_E_TIME_STAMP The timestamp signature and/or certificate could not be verified or is malformed.
```

This also explains why we cannot reproduce your problem. For loader onedal_thread and onedal_sequential we use safe `WinVerifyTrust` checks, most likely there are some problems.

We will try to deal with the problem in detail. As the news, we will definitely unsubscribe

**Comment by CHDev93** [comment]
Really appreciate you looking into this @PetrovKP . The forum you linked to seems to indicate this is something specific to my Windows setup. My takeaway was that there wasn't anything the library could do about this but some sort of patching I need on my machine. Is that your understanding as well?

**Comment by PetrovKP** [comment]
@CHDev93 DLL of oneDAL cannot verify the certificate on your system. Can you check the certificate properties to see if this is the problem?
![image](https://user-images.githubusercontent.com/17429885/107682852-ef049c00-6cb1-11eb-9f71-fac9440197c8.png)

**Comment by chaaland** [comment]
@PetrovKP I have a very similar "Digital Signature Details" page except the signing time is "12 November 2020 10:43:51". Specifically there is no day of the week and the date is order day, month, year rather than month day year. When I click "View Certificate", the general tab shows "The timestamp signature and/or certificate could not be verified or is malformed". The date certainly looks valid to me unless it's really not happy about european date formatting.

So this is certainly the problem as you mention, though I'm not familiar enough with these things to know how to proceed.

For reference the Valid from field shows 31/08/2020 to 17/06/2021. So again in day month year format

**Comment by PetrovKP** [comment, marked as answer]
@chaaland Unfortunately, it's hard for me to reproduce.  Updating certificates at the root can potentially help:

1. Download https://www.digicert.com/util/ digicert certificate utility
2. Run the utility and from the menu select “tools” and then AutoRootUpdate.
3. Change the setting to ‘Disabled’ . (Yes ‘Disabled’).

[Reference](https://community.logmein.com/t5/GoToAssist-Remote-Support/GoTo-Opener-Error-quot-Timestamp-signature-or-certificate-could/m-p/254105#M6375)

**Comment by CHDev93** [comment]
@PetrovKP managed to fixed this with your comments regarding the root certificate update. Thanks for your help!

Regarding the init.py not working for `virtualenv` on windows, should I create a separate issue?

**Comment by PetrovKP** [comment]
@CHDev93 Great news!

Today we have merged the [fix ](https://github.com/IntelPython/daal4py/pull/488)for the virtualenv issue.
